src/datasets/utils/py_utils.py

```diff
@@ -1330,8 +1330,33 @@ def _write_generator_to_queue(queue: queue.Queue, func: Callable[..., Iterable[Y
     return i
 
 
+class ProcessPool(multiprocess.pool.Pool):
+    """
+    A multiprocess.pool.Pool implementation that keeps track of child process' PIDs,
+    and can detect if a child process has been restarted.
+    """
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self._last_pool_pids = self._get_current_pool_pids()
+        self._has_restarted_subprocess = False
+
+    def _get_current_pool_pids(self) -> Set[int]:
+        return {f.pid for f in self._pool}
+
+    def has_restarted_subprocess(self) -> bool:
+        if self._has_restarted_subprocess:
+            # If the pool ever restarted a subprocess,
+            # we don't check the PIDs again.
+            return True
+        current_pids = self._get_current_pool_pids()
+        self._has_restarted_subprocess = current_pids != self._last_pool_pids
+        self._last_pool_pids = current_pids
+        return self._has_restarted_subprocess
+
+
 def iflatmap_unordered(
-    pool: Union[multiprocessing.pool.Pool, multiprocess.pool.Pool],
+    pool: ProcessPool,
```

Comment:
I think we still need to keep `Union[multiprocessing.pool.Pool, multiprocess.pool.Pool]` support in `iflatmap_unordered`